biopet-tools/src/main/scala/nl/lumc/sasc/biopet/tools/GtfToRefflat.scala

```diff
@@ -0,0 +1,141 @@
+    } text "Reference file"
+  }
+
+  def main(args: Array[String]): Unit = {
+    val argsParser = new OptParser
+    val cmdArgs
+      : Args = argsParser.parse(args, Args()) getOrElse (throw new IllegalArgumentException)
+
+    logger.info("Start")
+    gtfToRefflat(cmdArgs.gtfFile, cmdArgs.refFlat, cmdArgs.referenceFasta)
+    logger.info("Done")
+  }
+
+  def gtfToRefflat(gtfFile: File, refflatFile: File, referenceFasta: Option[File] = None): Unit = {
+    val reader = Source.fromFile(gtfFile)
+
+    val featureBuffer: mutable.Map[String, Int] = mutable.Map()
+
+    val referenceDict = referenceFasta.map(file => FastaUtils.getCachedDict(file))
+
+    logger.info("Readering gtf file")
+    val gtfLines = reader
+      .getLines()
+      .filter(!_.startsWith("#"))
+      .map(Feature.fromLine)
+      .map { feature =>
+        referenceDict.foreach(
+          dict =>
+            require(dict.getSequence(feature.contig) != null,
+                    s"Contig '${feature.contig}' does not exist on reference"))
+        featureBuffer += feature.feature -> (featureBuffer.getOrElse(feature.feature, 0) + 1)
+        feature
+      }
+      .toArray
+    logger.info("Readering gtf file - Done")
+
+    val gtfGenes = gtfLines.flatMap(_.attributes.get("gene_id")).toList.distinct
+    logger.info(s"${gtfGenes.size} genes found")
+
+    val gtfContent = gtfLines.toList
+      .groupBy(_.attributes.get("gene_id"))
+    logger.info("Readering gtf file - Done")
+
+    val genes = for ((geneId, gtfFeatures) <- gtfContent if geneId.isDefined) yield {
+      val gtfGene = gtfFeatures.find(_.feature == "gene").get
+      val gtfTranscripts = gtfFeatures.groupBy(_.attributes.get("transcript_id"))
+      val transcripts =
+        for ((transcriptId, features) <- gtfTranscripts if transcriptId.isDefined) yield {
+          val groupedFeatures = features.groupBy(_.feature)
+          val exons = groupedFeatures
+            .getOrElse("exon", Nil)
+            .sortBy(_.start)
+            .map(x => Exon(x.start, x.end))
+          val gtfTranscript = groupedFeatures("transcript").head
+          val cdsFeatures = groupedFeatures.get("CDS").map(_.flatMap(x => List(x.start, x.end)))
+          val startFeatures =
+            groupedFeatures.get("start_codon").map(_.flatMap(x => List(x.start, x.end)))
+          val stopFeatures =
+            groupedFeatures.get("stop_codon").map(_.flatMap(x => List(x.start, x.end)))
+          val codingLocations = startFeatures
+            .getOrElse(cdsFeatures.getOrElse(Nil)) ::: stopFeatures.getOrElse(
+            cdsFeatures.getOrElse(Nil))
+          val codingStart =
+            if (codingLocations.isEmpty) None
+            else if (gtfGene.strand.get) Some(codingLocations.min - 1)
+            else Some(codingLocations.min - 1)
+          val codingEnd =
+            if (codingLocations.isEmpty) None
+            else if (gtfGene.strand.get) Some(codingLocations.max)
+            else Some(codingLocations.max)
```

Comment:
Good question. This is because in a gtf file there are 2 ways of defining coding regions. They can use the feature `CDS` or a combination of `start_codon` and `stop_codon`. In annotation files this can be used both but not consistently :(.